File: exercises/java/src/main/java/cards/AnimalDeck.java
package cards;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class AnimalDeck implements Deck {

    private List<AnimalCard> cards;

    AnimalDeck() {
        cards = new ArrayList<>();
        for (Animal animal : Animal.values()) {
            cards.add(new AnimalCard(animal));
            cards.add(new AnimalCard(animal));
        }
    }

    public void shuffle() {
        Collections.shuffle(cards);
    }

    public String[] getCards() {
        String[] result = new String[cards.size()];
        for (int i = 0; i < cards.size(); i++) {
            AnimalCard card = cards.get(i);
            result[i] = card.toString();
        }
        return result;
    }

    public AnimalCard deal() {
        return cards.remove(0);
    }
}

File: exercises/kotlin/src/main/kotlin/cards/AnimalDeck.kt
package cards

internal class AnimalDeck {
    private val cards: MutableList<AnimalCard?>

    init {
        cards = ArrayList()
        for (animal in Animal.values()) {
            cards.add(AnimalCard(animal))
            cards.add(AnimalCard(animal))
        }
    }

    fun shuffle() {
        cards.shuffle()
    }

    fun getCards(): MutableList<String> {
        val result = mutableListOf<String>()
        for (i in cards.indices) {
            val card = cards[i]
            result.add(card.toString())
        }
        return result
    }

    fun deal(): AnimalCard? {
        return cards.removeAt(0)
    }
}